### DISCUSSION TITLE
TinyUF2 in pico-ice-sdk

### DISCUSSION THREAD

**Discussion by josuah** [adafruit/tinyuf2#272] [Show and tell]
Hello, and thank you for this open-source implementation of TinyUF2!

We are using it as part of an SDK, to provide UF2 flashing for an iCE40 FPGA board.
An RP2040 MCU is handling USB and connected to the FPGA (to start and stop it, i/o with it...) and its flash (for programming it).
The TinyUF2 bootloader have been reduced to a "[smol](https://www.youtube.com/watch?v=OkIzLycV_R0)" library that is now part of the pico-ice-sdk.

https://github.com/tinyvision-ai-inc/pico-ice-sdk/tree/main/src/tinyuf2

**Comment by josuah** [comment]
To be sure, is the license compatible with integration in the pico-ice-sdk?

**Reply by hathach** [reply]
since `pico-ice-sdk` is MIT license, and TinyUF2 is also MIT. I think we are all good.

**Reply by josuah** [reply]
Thank you for the confirmation! I will make sure to clean-up the LICENSE file to specify the origin of the various components involved after the project is completed.

**Comment by josuah** [comment]
The way it turned out, the TinyUF2 library we came-up with from your implementation could eventually be integrated directly into TinyUSB, just like it was done for DFU.

The TinyUF2 code you implemented is really portable, helping a lot with this.

Does it make sense?

**Reply by josuah** [reply]
I kept the sources as close as possible from your original work, to facilitate upgrades on the pico-ice-sdk, but if modifying them slightly, it could be reduced a bit further to only include what would be needed for TinyUSB integration.

**Reply by hathach** [reply]
can you elaborate "TinyUSB integration", do you mean making PR back to this repo (TinyUF2). TinyUSB is a different project.

**Reply by josuah** [reply]
This should have been a comment to add on the TinyUSB repository indeed.

What I was suggesting was taking the TinyUF2 code, shrink it down a bit.
Then turn it in a Tiny*USB* feature, so that TinyUSB can provide a UF2 programming interface, just like it provides a DFU programming interface.

This would be the opportunity for many device to get an UF2 interface easily configured.

I do not have time to try this *right now*, but when/if I do, I will try and let you know.

**Reply by hathach** [reply]
I don't think this should be part of TinyUSB, since TinyUSB focus on usb protocol and class driver, while UF2 is application-defined base on mass storage. UF2 programming should be taken care by this TinyUF2 project.

**Reply by gsteiert** [reply]
I see value in having UF2 as a library that could be included in other projects.  It might be better in its own repo so people can include it when needed, but it could also be a nice nice MSC example in TinyUSB.

**Reply by josuah** [reply]
> I don't think this should be part of TinyUSB, since TinyUSB focus on usb protocol and class driver, while UF2 is application-defined base on mass storage. UF2 programming should be taken care by this TinyUF2 project.

This makes sense! Thank you for your answer.
TinyUSB really taught me how the USB protocol works, and I like how it is built.
Thank you for that!

**Reply by josuah** [reply]
> I see value in having UF2 as a library that could be included in other projects. It might be better in its own repo so people can include it when needed, but it could also be a nice nice MSC example in TinyUSB.

Very good suggestion. Nothing prevents me to make it a separate independent repository.

**Reply by josuah** [reply]
@gsteiert 

> I see value in having UF2 as a library that could be included in other projects. It might be better in its own repo so people can include it when needed, but it could also be a nice nice MSC example in TinyUSB.

In the end, instead of forking TinyUF2, it is possible to keep it as a submodule. Only main.c needs to be replaced, and a custom back-end needs to be added.

This is the same approach that is taken with MicroPython: either define your board upstream either define your boards locally with MicroPython as a submodule.

**Reply by hathach** [reply]
yeah, I think it is possible, you only need to maintain your own tinyuf2 target in cmake file. Though be careful not to recursive submodule init tinyuf2 since it has other mcu drivers as part of its submodule list.

**Reply by josuah** [reply]
I confirm, this works very well!
https://github.com/tinyvision-ai-inc/pico-ice-sdk/blob/develop/CMakeLists.txt#L56-L59

I needed to copy then modify main.c to allow to cohabitate with the rest of the firmware, and declare an MSC interface in the existing usb_descriptors.c (the normal TinyUSB way).

This appears to work well! :)

It is used to program an FPGA and these UF2 toolbox was developed for this occasion: https://github.com/tinyvision-ai-inc/uf2-utils

Thank you for the advises!

**Reply by hathach** [reply]
glad that works for you. How differences is your main.c ? If not too much, we can actually make pico-ice as part of tinyuf2, maybe under port/rp2040_ice. This may come handy if there is more boards with pico + ice combo in the future ?

**Reply by josuah** [reply]
Apologies for the long silence, I was on the leave.

Reviewing again, the only change needed was to [remove `main()`](https://github.com/tinyvision-ai-inc/pico-ice-sdk/blob/develop/src/tinyuf2_main.c) (because there is another main for the user application running at the same time) and adding a [board definition file](https://github.com/tinyvision-ai-inc/pico-ice-sdk/blob/develop/src/tinyuf2_board.c).

If everything but `main()` is moved to a separate file, then it would be possible to use TinyUF2 as a library.

There are other dependencies from the pico-ice-sdk:
- `ice_spi.c` - bit-banged SPI because of wrong pinout of early version, but now it is good to convert it to hardware SPI,
- `ice_flash.c` - flash I/O commands to program it.
- `pico_ice.h` - RP2040 board definition that will be upstreamed in the future.

Two ways to do it I can think of:

- using the pico-ice-sdk to provide a maintained version of these small functions
- we provide an implementation of these functions directly into the board definition.

> if there is more boards with pico + ice combo in the future

There are a few hobby projects already around, but do not seems largely produced for now... Maybe more soon!

**Reply by hathach** [reply]
no problems, we are all busy. I think either having low level flash as local files of tinyuf2 or including pico-ice-sdk as submodule are OK. We don't have to get it right at the first time, we only need to make it convenience for us (maintainer) and other maker with same flashing (probably with different led, flash device etc ..) currently ice-pico is the only variant so there should be no problems. We could update/upgrade it later when needed. Feel free to submit PR whenever you have time to do so, I am sure will help to review it.

**Reply by josuah** [reply]
Many thanks for your openness.
I will do my best!
